src/python/python-proto/python_proto/api/plugin_sdk/analyzers/v1beta1/messages.py

```diff
@@ -0,0 +1,350 @@
+    src_uid: grapl_common_msgs.Uid
+    dst_uid: grapl_common_msgs.Uid
+    forward_edge_name: grapl_common_msgs.EdgeName
+    reverse_edge_name: grapl_common_msgs.EdgeName
+
+    _proto_cls = proto.EdgeUpdate
+
+    @classmethod
+    def from_proto(
+        cls,
+        proto_value: proto.EdgeUpdate,
+    ) -> EdgeUpdate:
+        return cls(
+            src_uid=grapl_common_msgs.Uid.from_proto(proto_value.src_uid),
+            dst_uid=grapl_common_msgs.Uid.from_proto(proto_value.dst_uid),
+            forward_edge_name=grapl_common_msgs.EdgeName.from_proto(
+                proto_value.forward_edge_name
+            ),
+            reverse_edge_name=grapl_common_msgs.EdgeName.from_proto(
+                proto_value.reverse_edge_name
+            ),
+        )
+
+    def into_proto(self) -> proto.EdgeUpdate:
+        proto_value = self.new_proto()
+        proto_value.src_uid.CopyFrom(self.src_uid.into_proto())
+        proto_value.dst_uid.CopyFrom(self.dst_uid.into_proto())
+        proto_value.forward_edge_name.CopyFrom(self.forward_edge_name.into_proto())
+        proto_value.reverse_edge_name.CopyFrom(self.reverse_edge_name.into_proto())
+        return proto_value
+
+
+UpdateInner = (
+    StringPropertyUpdate | UInt64PropertyUpdate | Int64PropertyUpdate | EdgeUpdate
+)
+
+
+@dataclasses.dataclass(frozen=True, slots=True)
+class Update(SerDe[proto.Update]):
+    inner: UpdateInner
+
+    _proto_cls = proto.Update
+
+    @classmethod
+    def from_proto(cls, proto_value: proto.Update) -> Update:
+        field_name = proto_value.WhichOneof("inner")
+        assert field_name is not None
+
+        match field_name:
+            case "string_property":
+                return cls(
+                    inner=StringPropertyUpdate.from_proto(proto_value.string_property)
+                )
+            case "uint64_property":
+                return cls(
+                    inner=UInt64PropertyUpdate.from_proto(proto_value.uint64_property)
+                )
+            case "int64_property":
+                return cls(
+                    inner=Int64PropertyUpdate.from_proto(proto_value.int64_property)
+                )
+            case "edge":
+                return cls(inner=EdgeUpdate.from_proto(proto_value.edge))
+
+        raise Exception(f"Unknown variant: {field_name}")
+
+    def into_proto(self) -> proto.Update:
+        msg = self.new_proto()
+        match self.inner:
+            case StringPropertyUpdate() as inner:
+                msg.string_property.CopyFrom(inner.into_proto())
+            case UInt64PropertyUpdate() as inner:
+                msg.uint64_property.CopyFrom(inner.into_proto())
+            case Int64PropertyUpdate() as inner:
+                msg.int64_property.CopyFrom(inner.into_proto())
+            case EdgeUpdate() as inner:
+                msg.edge.CopyFrom(inner.into_proto())
+            case _:
+                raise Exception(f"Unknown variant: {self.inner}")
```

Comment:
Isn't this exhaustive?

Reply:
https://peps.python.org/pep-0622/#check-exhaustiveness-at-runtime

i wouldn't expect this to ever happen, in your PR it's `assert_never`

Reply:
ya assert_never is mypy time tho, nbd